### PR TITLE
Initial hack to get dask distributed working

### PR DESCRIPTION
**Pull request by shoyer**
WIP WIP WIP

CC
@mrocklin
@pwolfram 

xref #798

### CODE REVIEW

**Comment by mrocklin** [comment]
It occurs to me after reading your DataStore API issue that we could also just implement getstate/setstate methods on all of the wrapping classes (like LazilyIndexedArray) that thought about how to serialize their contained arrays.  You wouldn't have to use dask's mechanism.  

You would still have to figure out how to effectively serialize and deserialize any contained object, which is the same problem we're solving at https://github.com/dask/distributed/pull/620 , but could stick to the pickle protocol if you wanted to stay with more general and established protocols.  You are reaching down a little bit in the object tree further than was intended, but that might be a more palatable long term solution for xarray?

**Comment by mrocklin** [comment]
Not that I'm recommending this, just that it is another option.

File: xarray/conventions.py [hunk 1]
```diff
@@ -470,6 +476,27 @@ def __getitem__(self, key):
...
+    header['__char_to_string_array__array_type'] = header.get('type')
+    return header, frames
+
+
+def deserialize_char_to_string_array(header, frames):
+    type_ = header.pop('__char_to_string_array__array_type', None)
+    if type_ is not None:
+        header['type'] = type_
+    array = distributed.protocol.deserialize(header, frames)
+    return CharToStringArray(array)
```
Comment (shoyer) [inline xarray/conventions.py]: That would be fine, but these classes (other than `LazilyIndexedArray`) are not going to be layered multiple times. So I wouldn't worry too much about that until we settle on a general solution.

**Comment by shoyer** [comment]
> It occurs to me after reading your DataStore API issue that we could also just implement getstate/setstate methods on all of the wrapping classes (like LazilyIndexedArray) that thought about how to serialize their contained arrays. You wouldn't have to use dask's mechanism.

If I understand correctly, I could even just go to the bottom level, and make `NetCDF4ArrayWrapper` pickle-able, using a similar strategy to the one in the dask-distributed PR.

This does seem a little more palatable in the long term, especially if we build some more general machinery that DataStores can plug into (added into https://github.com/pydata/xarray/pull/1087)

**Comment by shoyer** [comment]
Closing in favor of https://github.com/pydata/xarray/pull/1095

**Pull request closed without merging**